feat(mdx-loader): the table-of-contents should display toc/headings of imported MDX partials

anatolykopyl opened the pull request:
Note: @anatolykopyl started this PR, and then @slorber took it over.

## Motivation

Fix https://github.com/facebook/docusaurus/issues/3915

It is now possible to import MDX partials and have headings of the partial appear in the TOC.

We'll consider this as a "new feature" instead of a bug fix, considering it remains an important behavior change, and an improvement over stock MDX. We'll release this in v3.2  instead of v3.1.1

---

Consider the following: 

```mdx
// _partial.mdx

## Partial heading
```

```mdx
// doc.mdx

# Doc

## Doc heading

import Partial from "_partial.mdx"

<Partial/>
```

This doc now renders a TOC with 2 headings, while previously "Partial heading" didn't appear in the TOC.

In practice, the doc is now compiled as:

```mdx
// doc.mdx
import Partial, 
  {toc as __tocPartial} from "_partial.mdx"

# Doc

## Doc heading

export const toc = [
  {
    "value": "Doc heading",
    "id": "doc-heading",
    "level": 2
  }, 
  ...__tocPartial,
]
```

Where `__tocPartial` contains `{"value": "Partial heading","id": "partial-heading","level": 2}`

This makes it possible to have headings working for "deeply-nested-partials" where one partial import another.

**Notable change**: if you write `export const toc = ...` in your MDX docs, we now never override this value anymore. Previously we did, but only when there were headings. We consider this as a bug fix and not a breaking change.

## Test Plan

Unit tests + dogfood

### Test links

https://deploy-preview-9684--docusaurus-2.netlify.app/tests/pages/partials-tests

## Related issues/PRs

This PR adapts a similar fix from @dimaMachina made in Nextra to Docusaurus: https://github.com/shuding/nextra/pull/2199

anatolykopyl committed:
Headings from partials appear in ToC

anatolykopyl committed:
Fixed generated mdxjsEsm node value and added test page

anatolykopyl commented:
I consider this PR done.
@slorber, could you please check it out? Any tips regarding how this can be improved are appreciated! I'd be glad to make changes to this PR if necessary.

slorber commented on packages/docusaurus-mdx-loader/src/remark/utils/index.ts:
```diff
@@ -154,3 +155,32 @@ export function assetRequireAttributeValue(
   };
 }
+
+/**
+ * Similar to stringify-object, but keeps spread operators,
+ * instead of turning them into strings.
+ * @param objects
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function constructArrayString(objects: any[]): string {
+  let result = '[';
+
+  for (const obj of objects) {
+    if (typeof obj === 'string') {
+      result = `${result}\n\t${obj},`;
+    } else {
+      result = `${result}\n\t${stringifyObject(obj).replace(/\n/g, '\n\t')},`;
+    }
+  }
+
+  // Remove trailing coma
+  result = result.replace(/,$/, '');
+  // Close array on same line if it is empty
+  if (result === '[') {
+    result += ']';
+  } else {
+    result += '\n]';
+  }
+
+  return result;
+}
```
wonder if we shouldn't use something cleaner for that instead of string concat.

For example: https://github.com/syntax-tree/estree-util-to-js

anatolykopyl replied:
I removed this whole function as per @wooorm's suggestion. But in case we need `value` to be set, this can be done with [astring](https://github.com/davidbonnet/astring), which seems to do the job well.

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -124,9 +178,21 @@ export default plugin;
 async function createExportNode(name: string, object: any): Promise<MdxjsEsm> {
```
I know it was like that before, but that object shouldn't be of type `any`. It should have previously been `TOCItem[]`, and now support your additional types

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,21 +93,75 @@ const plugin: Plugin = function plugin(
+            continue;
+          }
+          const {length} = Object.keys(partialComponentToHeadingsName);
+          const exportAsName = `${name}${length}`;
```
I'd rather prefix it with `__` like nextra does

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,21 +93,75 @@ const plugin: Plugin = function plugin(
+        }
+        const headingsName = partialComponentToHeadingsName[nodeName];
+        if (headingsName) {
+          headings.push(`...${headingsName}`);
```
Not super fan of using this "string-based data structure": you add "..." to later remove this prefix. Maybe use some kind of object instead?

anatolykopyl replied:
Came up with a new type called `NestedTOC`, that is used like this:
```typescript
type NestedTOC = {
  readonly nested: true;
  readonly name: string;
};

const headings: (TOCItem | NestedTOC)[] = [];
``` 

This is because I need to store both types in one array, but at some point I have to process them differently.

Can you give your thoughts on this?

slorber replied:
I don't really have a better name and it's only a local type so that's good enough to me 👍

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/__tests__/__snapshots__/index.test.ts.snap:
```diff
@@ -236,3 +236,23 @@ Some content here
+		id: 'thanos',
+		level: 2
+	},
+	...toc0
```
toc0 is not imported here, is this expected?

anatolykopyl replied:
Seems like we're going down the road of only modifying the AST and not the actual js, so this is expected.

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,21 +93,75 @@ const plugin: Plugin = function plugin(
-        id: child.data!.id!,
-        level: child.depth,
-      });
+      if (child.type === 'mdxjsEsm') {
```
I know the original code was already low-level, but I'd appreciate if we could make it more readable and high-level. I mean the intent of this algorithm should be clear without spending time trying to understand the implementation details. Currently it's a bit hard to see the big picture of it. I would extract things in smaller functions for example.

anatolykopyl replied:
Moved the visitor for each type of node into its own function. This is a bit more readable now, but I can try to do some further reorganisation if needed.

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -124,9 +178,21 @@ export default plugin;
 
+  const tocObject = object.map((heading: TOCItem | string) => {
+    if (typeof heading === 'string') {
+      const argumentName = heading.replace('...', '');
```
I'd prefer to represent this thing as an object rather than a ... prefixed string

slorber commented on website/_dogfooding/_pages tests/partials-tests.mdx:
```diff
@@ -0,0 +1,12 @@
+import PagePartial from './_pagePartial.mdx';
```
our dogfood section is messy currently.

I'd prefer if you created this in `docusaurus/website/_dogfooding/_docs tests/toc-partials` and added much more complex cases such as "deeply-nested partials" (one partial importing another)

slorber reviewed (changes requested):
That looks like good start thanks 👍 

I'm not 100% fan of the implementation yet but overall the solution looks like what we want to adopt.

I'd like to have much more unit tests, testing more complex cases, such as using the same partial twice in a page, or importing the same partial using different names.

That would also be interesting to dogfood with more complex cases, such as a deeply nested tree of partials (having partials importing other partials)

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -124,9 +178,21 @@ export default plugin;
   return {
     type: 'mdxjsEsm',
-    value: `export const ${name} = ${stringifyObject(object)}`,
+    value: `export const ${name} = ${constructArrayString(object)}`,
```
Little question for @wooorm or @dimaMachina

Could you explain when to use `mdxjsEsm.value` and when to use `mdxjsEsm.data.estree.body`? 

It's not clear to me which one will be used for what exactly.

On the Nextra PR, it seems they don't use `mdxjsEsm.value` and yet it works. However if I remove it on our remark plugin the toc is not exported anymore.

---

Original Nextra code:

https://github.com/shuding/nextra/pull/2199/files

<img width="683" alt="CleanShot 2024-01-12 at 22 40 56@2x" src="https://github.com/facebook/docusaurus/assets/749374/5e1f8262-77b2-4da5-a13f-e00cd238f4fe">

wooorm replied:
It’s the same thing as when strings of raw HTML are supported and when ASTs of nodes are supported. The string is used by tools that don’t understand JavaScript ASTs (remark). The AST is used by tools that don’t understand strings of JavaScript (recma).

In this case, you are not generating MDX files—you are not compiling to markdown. So you don’t need `value`.
You are generating JavaScript. So you need JavaScript ASTs.

> However if I remove it on our remark plugin the toc is not exported anymore.

I don’t know what these words practically mean. You are explaining a bug that I very much doubt would happen.

---

> [src/remark/toc/utils.ts](https://github.com/facebook/docusaurus/pull/9684#diff-50ad45314b03426612bb4a598f8adf0cc099f4bfc8674528aa358ed01aaaacaf)

I recommend dropping `isImport`, `isExport`, and `hasExports`. The types aren’t very good, as you can see from the casts. Take a look at the places these utilities are used, and which types TypeScript knows of there. TS is already much smarter there.

anatolykopyl replied:
From my testing, removing value doesn't break the generation of ToC. But I'm concerned with the possibility of another plugin relying on `value`. So I feel like it would be right to keep it. At least since it's a required property in `MdxjsEsm`.

wooorm replied:
I recommend an empty string `''` for now instead of pulling in `astring`. Practically there are no plugins that do, or should, use the string of JavaScript, in turning markdown into javascript.
You can always add it later when a plugin like that appears!

slorber replied:
Thanks for the answer @wooorm I understand better now.

The problem was probably only in our unit test snapshots since we serialize back to Markdown:

```tsx
const processFixture = async (name: string) => {
  const {remark} = await import('remark');
  const {default: gfm} = await import('remark-gfm');
  const {default: mdx} = await import('remark-mdx');

  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  const result = await remark()
    .use(headings)
    .use(gfm)
    .use(mdx)
    .use(plugin)
    .process(file);

  return result.value;
};
```

This makes the snapshots more convenient to review than using AST or JSX IMHO, so I'd like to keep this if possible.

But as far as I understand it keeping this format would require using astring?

Do you see any solution to this problem? 

I was thinking we could create an astring remark plugin generating values just during tests for example?

slorber replied:
Can confirm it works nicely and helps keeping readable snapshots:

```js
visit(root, (child) => {
        if (child.type === 'mdxjsEsm') {
          const node = child as MdxjsEsm;
          node.value = node.data?.estree
            ? generate(node.data.estree)
            : node.value;
        }
});
```

<img width="495" alt="CleanShot 2024-01-18 at 18 13 02@2x" src="https://github.com/facebook/docusaurus/assets/749374/f8a9834c-864e-4824-975e-8dffc1d49a3c">

wooorm replied:
> The problem was probably only in our unit test snapshots since we serialize back to Markdown:

Here you specifically care about JavaScript; not markdown.
So use MDX to compile to JavaScript. Snapshot that. Much easier to review than a fake string that never actually runs!

`compile(input, {remarkPlugins, rehypePlugins})`

slorber replied:
👍 makes sense, will change the snapshot format

wooorm replied:
btw don’t treat my comments as blockers — just random ideas about how I would do it!

anatolykopyl commented:
Thanks for the detailed feedback! I'll be working on improving this PR in the following weeks.

anatolykopyl committed:
Cleanup code and use astring instead of string concat

anatolykopyl committed:
Js now generates with import, not just the AST

Josh-Cena commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,66 +93,158 @@ const plugin: Plugin = function plugin(
     const {visit} = await import('unist-util-visit');
 
-    const headings: TOCItem[] = [];
+    const partialComponentToHeadingsName: {[key: string]: string} =
+      Object.create(null);
```
I suggest using a map

Josh-Cena reviewed (commented):
I'm quite happy with this implementation; exactly how I pictured it.

wooorm commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,66 +93,158 @@ const plugin: Plugin = function plugin(
+      }
+    }
+
+    visit(root, ['heading', 'mdxjsEsm', 'mdxJsxFlowElement'], (child) => {
```
```suggestion
    visit(root, (child) => {
```

This test is useless if you are already testing.
And the types are better if you don’t use arrays. TypeScript doesn’t understand the types of arrays.

anatolykopyl replied:
Removed the array, but can't figure out how to remove the casts. Checking for `child.type` has no effect on the type of `child` inside the `if` statement.

Did I not fully understand you?

wooorm replied:
> Checking for child.type has no effect on the type of child inside the if statement.

It should, if `root` is typed as a `Root`.
It looks like the types here don’t work because `Plugin` is `any`.
Typically, plugins have a type such as `Plugin<[], Root>` would type it.
You should alternatively be able to type `root` on L92 (`root: Root`). Or on L89: `Transformer<Root>`.

slorber replied:
Can confirm I still see casting is needed after the change

<img width="1009" alt="CleanShot 2024-01-18 at 16 23 09@2x" src="https://github.com/facebook/docusaurus/assets/749374/50d20656-6561-4e61-85aa-67ee3915585b">

anatolykopyl committed:
Update packages/docusaurus-mdx-loader/src/remark/toc/index.ts

Co-authored-by: Titus <tituswormer@gmail.com>

anatolykopyl committed:
Added dogfooding and changed unit tests

anatolykopyl committed:
Moved types into the file where they are used

anatolykopyl commented:
Moved the dogfooding into [`_docs tests`](https://deploy-preview-9684--docusaurus-2.netlify.app/tests/docs/toc-partials/). 
Added a deeply nested partial and a partial with props in the heading. Last one is not working, the heading shows the text passed to the prop as expected, but the ToC displays just the text `props.propname`. Quite interesting, no idea how this can possibly be fixed.

Also since I changed the plugin to not generate any js anymore, only ASTs, the existing unit tests for the ToC plugin became useless, as they checked generated mdx. I changed the unit tests to output ASTs, so the work of the plugin is visible.

wooorm commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -93,21 +94,82 @@ const plugin: Plugin = function plugin(
+      }
+    }
+
+    visit(root, (child: Node) => {
```
Remove `: Node` here, then you can also remove the other `as` casts.
The types for the ASTs are already quite smart, casts like `: Node` and `as Heading` should not be needed

slorber replied:
For me this doesn't work:

```tsx
    visit(root, (child) => {
      if (child.type === 'heading') {
        const heading: Heading = child
      } 
    });
```

How is this supposed to work?

Maybe we should ask TS to load some declaration merging typedefs containing a registry of types?

wooorm reviewed (commented):
> Last one is not working, the heading shows the text passed to the prop as expected, but the ToC displays just the text props.propname.

What output do you want, what would be “working”?

The code you use uses `toValue`: https://github.com/anatolykopyl/docusaurus/blob/1fdd6af8c79e7765ff4d7b4731e605d0d95c1316/packages/docusaurus-mdx-loader/src/remark/utils/index.ts#L76C17-L76C24.
Which calls `toString()` on an expression (`{props.propname}`): https://github.com/anatolykopyl/docusaurus/blob/1fdd6af8c79e7765ff4d7b4731e605d0d95c1316/packages/docusaurus-mdx-loader/src/remark/utils/index.ts#L99
Which returns `node.value`, which is `props.propname`.

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/__tests__/index.test.ts:
```diff
@@ -17,14 +17,7 @@ const processFixture = async (name: string) => {
-    .process(file);
-
-  return result.value;
+  return remark().use(headings).use(gfm).use(mdx).use(plugin).parse(file);
```
This change makes snapshots much more verbose and harder to review.

Also, since snapshot format has changed, we now can't easily know if there is a TOC regression since this also updates all former tests snapshots. Keeping the former test snapshots intact is what guarantees your logic does not introduce any regression.

If you really need to capture AST in snapshots, please explain why, and do this only for the newly added tests so that we can still be sure there's no toc regression in this PR.

slorber commented on yarn.lock:
```diff
@@ -4423,7 +4423,7 @@ astral-regex@^2.0.0:
   integrity sha512-Z7tMw1ytTXt5jqMcOP+OQteU1VuNK9Y02uuJtKQ1Sv69jXQKKg5cibLwGJow8yzZP+eAc18EmLGPal0bp36rvQ==
 
-astring@^1.8.0:
+astring@^1.8.0, astring@^1.8.6:
```
```suggestion
astring@^1.8.0:
```

unnecessary change?

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -7,41 +7,42 @@
+// TODO upgrade to TS 5.3
+// See https://github.com/microsoft/TypeScript/issues/49721#issuecomment-1517839391
+// import type {Plugin} from 'unified';
+type Plugin = any;
```
Was confused to see a diff on this. Move it at the top above TOCItem like it was before to remove the useless diff?

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -7,41 +7,42 @@
   readonly level: number;
 };
 
+type NestedTOC = {
```
What about "TOCSlice"?

This represents something we can spread into the original toc array

slorber commented on packages/docusaurus-mdx-loader/src/remark/toc/index.ts:
```diff
@@ -7,41 +7,42 @@
 };
 
+type NestedTOC = {
+  readonly nested: true;
```
I'd prefer to have `type: "item"` and `type: "slice"`

This way to discriminate the 2 types felt weird to me: `if ('nested' in heading)`

slorber replied:
We can't modify TOCItem because it's exported and used in theme. Will find another solution to keep things retro compatible

slorber commented on website/_dogfooding/_docs tests/toc-partials/_partial-with-props.mdx:
```diff
@@ -0,0 +1,5 @@
+## This is a partial markdown with props
+
+### Props in headings dont work: {props.heading}
```
Don't worry about that edge case, we don't support expressions in headings even in non-partial files atm. If we work on this it will be in another PR.

slorber reviewed (changes requested):
Thanks, that looks like it's working nicely.

The `{props.propname}` is not a problem to consider for now so you can remove that test.

I understand why you changed the test snapshots to MDAST although I don't like it. I'll try to come up with a solution and commit a thing to your PR soon.

slorber committed:
restore former TOC snapshot format

slorber committed:
TOC, revert usage to Markdown snapshots thanks to astring.generate(jsAst)

slorber committed:
better unit tests

slorber committed:
Add astring to devDependencies

slorber commented:
Thanks @anatolykopyl , looks like it's working nicely.

Almost ready, I'm just going to refactor some details and merge asap!

slorber committed:
typo

slorber committed:
typo

slorber committed:
typo

slorber committed:
Minor changes to site preprocessor

anatolykopyl committed:
Remove some unnecessary type casts

slorber committed:
better dogfooding case

slorber committed:
use MdxjsEsm type guard

slorber committed:
more refactors and function extractions

slorber committed:
more refactors

slorber committed:
more refactors + remove useless babel parser dependency

slorber committed:
refactor handling of toc export node

slorber committed:
refactor handling of toc export node

slorber committed:
update yarn lock

slorber committed:
refactor implementation, solve edge case where partial import is done after partial usage

slorber committed:
add missing typeguard

slorber committed:
minor naming refactors

slorber committed:
minor refactoring

slorber committed:
refactor TOC snapshots to use JS instead of MDX

slorber committed:
never override user-provided toc export

slorber committed:
add export toc at the end instead of after last import

slorber committed:
minor utils refactor

slorber committed:
rename tocSlice.name to tocSlide.importName

slorber commented:
Thanks for your initial port @anatolykopyl, now it's ready to merge 🎉 

---

Thanks for the Nextra inspired solution @dimaMachina

FYI a little detail in your implementation that does not work is when the import is after the usage. Probably not a big deal, but I fixed it in this PR:

```mdx
<Partial/>

import Partial from "partial";
```

That would be interesting to see if we could find a shared abstraction that benefit to Docusaurus, Nextra and others, but not sure it's possible atm considering the little differences here and there: we probably need to chat and design the whole thing.

slorber committed:
remove useless toc explicit exports

slorber committed:
remove useless toc explicit exports on website
